src/history.rs

```diff
@@ -499,34 +481,22 @@ impl HistoryImpl {
         self.loaded_old = true;
 
         let _profiler = TimeProfiler::new("load_old");
-        if let Some(filename) = history_filename(&self.name, L!("")) {
-            let Ok(mut file) = wopen_cloexec(&filename, OFlag::O_RDONLY, Mode::empty()) else {
-                return;
-            };
-
-            // Take a read lock to guard against someone else appending. This is released after
-            // getting the file's length. We will read the file after releasing the lock, but that's
-            // not a problem, because we never modify already written data. In short, the purpose of
-            // this lock is to ensure we don't see the file size change mid-update.
-            //
-            // We may fail to lock (e.g. on lockless NFS - see issue #685. In that case, we proceed
-            // as if it did not fail. The risk is that we may get an incomplete history item; this
-            // is unlikely because we only treat an item as valid if it has a terminating newline.
-            let locked = unsafe { Self::maybe_lock_file(&mut file, LOCK_SH) };
-            self.file_contents = HistoryFileContents::create(&mut file);
-            self.history_file_id = if self.file_contents.is_some() {
-                file_id_for_file(&file)
-            } else {
-                INVALID_FILE_ID
-            };
-            if locked {
-                unsafe {
-                    Self::unlock_file(&mut file);
+        if let Ok(history_path) = self.history_file_path() {
+            match lock_and_load(
+                &history_path,
+                &LOCK_HISTORY_FILE,
+                HistoryFileContents::create,
+            ) {
+                Ok((file_id, file_contents)) => {
+                    self.file_contents = Some(file_contents);
+                    self.history_file_id = file_id;
+                    let _profiler = TimeProfiler::new("populate_from_file_contents");
+                    self.populate_from_file_contents();
+                }
+                Err(e) => {
+                    FLOG!(history_file, "Error reading from history file:", e);
```

Comment:
notably this will log an error on the first run where the history file doesn't exist yet.
But that's fine, since this log is disabled by default. Seems reasonable.